server/src/main/java/org/elasticsearch/transport/Transport.java

```diff
@@ -231,14 +235,17 @@ long newRequestId() {
         /**
          * Removes and returns all {@link ResponseContext} instances that match the predicate
          */
-        public List<ResponseContext<? extends TransportResponse>> prune(Predicate<ResponseContext<? extends TransportResponse>> predicate) {
-            final List<ResponseContext<? extends TransportResponse>> holders = new ArrayList<>();
+        public Map<Long, ResponseContext<? extends TransportResponse>> prune(
+            Predicate<ResponseContext<? extends TransportResponse>> predicate
+        ) {
+            final Map<Long, ResponseContext<? extends TransportResponse>> holders = new HashMap<>();
```

Comment:
I wonder if maybe `ResponseContext` should know its `requestId`, avoiding the need for a map here? I think that might be useful in other logging too.

Reply:
Sure, we can do that. Applied to PR and made also ResponseContext a record.